File: desktop/core/pageMetrics.cpp
#include "pageMetrics.hpp"

namespace stencil::core {

  // Mirrors PAGE_SIZES in browser/js/config/constants.json.
  PageSize namedPageSize(const std::string& name) {
    if (name == "A3") return {29.7, 42.0};
    if (name == "A4") return {21.0, 29.7};
    return {0.0, 0.0};
  }

  // Port of DrawingApp.getPageDimensions: custom passes through; a named size
  // swaps to landscape when the image is wider than tall.
  PageSize pageDimensions(const std::string& name,
                          int canvasWidth, int canvasHeight,
                          double customWidth, double customHeight) {
    if (name == "custom") return {customWidth, customHeight};
    const PageSize ps = namedPageSize(name);
    if (canvasWidth > canvasHeight) return {ps.height, ps.width};
    return ps;
  }

  // Port of DrawingApp.pixelToPageCoords (raw part, pre-formula).
  Point pixelToPageRaw(double x, double y,
                       const PageSize& dims, int canvasWidth, int canvasHeight) {
    Point p;
    p.x = (canvasWidth  != 0) ? (dims.width  / canvasWidth)  * x : 0.0;
    p.y = (canvasHeight != 0) ? (dims.height / canvasHeight) * y : 0.0;
    return p;
  }

}

File: browser/js/core/parse/pageMetrics.js
import { core } from '../abi/stencilCore.js';
import constants from '../../config/constants.json' with { type: 'json' };
const { PAGE_SIZES } = constants;

// A wasm-parity surface (core/page/pageMetrics): the core owns the named-size table and the
// pixel→cm scaling when loaded; the JS is the reference and the fallback.

export const getPageDimensions = (app) => {
  const fn = core.op('pageDimensions');
  if (fn) {
    return fn(app.pageSize, app.canvas.width, app.canvas.height,
      app.customPageWidth, app.customPageHeight);
  }
  if (app.pageSize === 'custom') return { width: app.customPageWidth, height: app.customPageHeight };
  const ps = { ...PAGE_SIZES[app.pageSize] };
  if (app.canvas.width > app.canvas.height) return { width: ps.height, height: ps.width };
  return ps;
};

// What the f(x,y) names resolve to: the page in cm, the image in pixels — absent until one
// is open, so IMAGE_WIDTH / IMAGE_HEIGHT stay unknown — and the active display unit.
export const formulaContext = (app, ps = app.canvas ? getPageDimensions(app) : {}) => ({
  pageWidthCm: ps.width,
  pageHeightCm: ps.height,
  imageWidth: app.image ? app.canvas.width : undefined,
  imageHeight: app.image ? app.canvas.height : undefined,
  unit: app.unit
});

export const pixelToPageCoords = (app, x, y) => {
  const ps = getPageDimensions(app);
// formula.applyCtx itself already routes through the wasm parser (see FormulaEngine).
  const pixelToPageRaw = core.op('pixelToPageRaw');
  const raw = pixelToPageRaw
    ? pixelToPageRaw(x, y, ps, app.canvas.width, app.canvas.height)
    : { x: (ps.width / app.canvas.width) * x, y: (ps.height / app.canvas.height) * y };
// Both axes are in reach of either formula, so f(x) may read y and f(y) may read x.
  const ctx = { ...formulaContext(app, ps), x: raw.x, y: raw.y };
  return {
    x: app.formula.applyCtx(app.formulaX, 'x', raw.x, app.allowFormulas, ctx),
    y: app.formula.applyCtx(app.formulaY, 'y', raw.y, app.allowFormulas, ctx)
  };
};
